// trading-journal/src/hooks/useMonteCarlo.ts
// @apps/trading-journal/src/hooks/useMonteCarlo.ts

import { useState, useCallback, useRef, useEffect } from 'react';
import type { 
  MonteCarloRun, 
  MonteCarloSummary, 
  MonteCarloConfig, 
  MonteCarloHistoryItem 
} from '../types/monteCarlo'; 

import { saveHistory } from '../services/monteCarloService'; 

// 游눤 CORRE칂츾O 1: Use esta sintaxe para importar o construtor do Web Worker.
// Isso resolve o erro 'Module has no default export'.
import WorkerConstructor from '../workers/monteCarloWorker?worker'; 


export interface UseMonteCarloReturn {
  running: boolean;
  progress: number;
  summary: MonteCarloSummary | null;
  sampleRuns: MonteCarloRun[] | null;
  loadedHistoryId: string | null; // 游눤 NOVO: ID do hist칩rico carregado
  runSimulation: (config: MonteCarloConfig) => Promise<void>;
  stopSimulation: () => void;
  loadFromHistory: (item: MonteCarloHistoryItem | null) => void; 
}


export const useMonteCarlo = (): UseMonteCarloReturn => {
    
    // ESTADOS
    const [running, setRunning] = useState(false);
    const [progress, setProgress] = useState(0);
    const [summary, setSummary] = useState<MonteCarloSummary | null>(null);
    const [sampleRuns, setSampleRuns] = useState<MonteCarloRun[] | null>(null);
    const [loadedHistoryId, setLoadedHistoryId] = useState<string | null>(null); // 游눤 NOVO ESTADO

    const workerRef = useRef<Worker | null>(null); 
    const configRef = useRef<MonteCarloConfig | null>(null); 
    
    
    useEffect(() => {
        // 1. Inicializa o worker
        const worker = new WorkerConstructor(); // 游눤 Usa o construtor correto
        workerRef.current = worker;

        // 2. Listener de Mensagens
        worker.onmessage = (ev: MessageEvent) => {
            const msg = ev.data;
            if (msg.type === 'PROGRESS') {
                setProgress(msg.progress);
            } else if (msg.type === 'RESULT') {
                setSummary(msg.summary);
                setSampleRuns(msg.sampleRuns);
                setRunning(false);
                setProgress(100);

                // 3. Salva o hist칩rico (o refresh vir치 do MonteCarloPage)
                if (configRef.current) {
                    saveHistory(configRef.current, msg.summary, msg.sampleRuns);
                }
            } else if (msg.type === 'ERROR') {
                console.error("Worker Error:", msg.error);
                setRunning(false);
                setProgress(0);
            }
        };

        // 4. Cleanup
        return () => {
            if (workerRef.current) {
                workerRef.current.terminate();
            }
        };
    }, []); 

    // ... (stopSimulation permanece) ...
  // 游눤 Mantenha a ordem: Fun칞칫es de A칞칚o primeiro
    const stopSimulation = useCallback(() => {
        if (workerRef.current) {
            workerRef.current.terminate();
            // Recria o worker
            workerRef.current = new WorkerConstructor(); 
            // Re-adiciona o listener (se necess치rio, mas o useEffect j치 faz isso no unmount/mount)
        }
        setRunning(false);
    }, []); // N칚o depende de nada
    
    const runSimulation = useCallback(async (config: MonteCarloConfig) => {
        if (running || !workerRef.current) return;
        
        setRunning(true);
        setProgress(0);
        setSummary(null);
        setSampleRuns(null);
        setLoadedHistoryId(null); // Limpa o ID ao iniciar nova simula칞칚o
        
        configRef.current = config; 

        workerRef.current.postMessage({ 
            type: 'START', 
            config: config,
            empiricalR: undefined
        });
    }, [running]);

    
    // 游눤 CORRIGIDO: Aceita null para limpar a visualiza칞칚o. Define o ID do hist칩rico.
    const loadFromHistory = useCallback((item: MonteCarloHistoryItem | null) => {
        if (!item) {
             setSummary(null);
             setSampleRuns(null);
             setLoadedHistoryId(null); // Limpa o ID
             return;
        }
        setSummary(item.summary);
        setSampleRuns(item.sampleRuns || null); 
        setRunning(false);
        setProgress(100);
        setLoadedHistoryId(item.id); // Define o ID do hist칩rico
    }, []);

    return {
        running, progress, summary, sampleRuns, loadedHistoryId,
        runSimulation, stopSimulation, loadFromHistory,
    };
};
