// @blocknote/core@0.41.1/dist/blocks.js
import { Z as t, E as c, a as s, F as o, r as l, Q as r, ab as i, b as g, d as k, e as B, a7 as n, X as p, c as d, f as u, x as S, y as C, A as f, z as m, g as I, h as T, _ as L, j as P, k as E, l as _, n as h, o as D, q as H, s as b, w as A, B as M, C as O, H as N, I as V, K as x, L as F, M as G, N as y, P as v, D as W, G as Y, V as Q, R, T as w, a1 as U, $ as j, a5 as q, a4 as z, a9 as J, a3 as K, a2 as X, U as Z, a6 as $, m as ee, ac as ae, i as te, Y as ce, ae as se, ad as oe, t as le, u as re, v as ie, a8 as ge, a0 as ke, p as Be, aa as ne, O as pe, W as de, S as ue, J as Se } from "./BlockNoteSchema-CYRHak18.js";
export {
  t as BlockNoteSchema,
  c as EMPTY_CELL_HEIGHT,
  s as EMPTY_CELL_WIDTH,
  o as FILE_AUDIO_ICON_SVG,
  l as FILE_IMAGE_ICON_SVG,
  r as FILE_VIDEO_ICON_SVG,
  i as addDefaultPropsExternalHTML,
  g as audioParse,
  k as audioRender,
  B as audioToExternalHTML,
  n as blockHasType,
  p as checkPageBreakBlocksInSchema,
  d as createAudioBlockConfig,
  u as createAudioBlockSpec,
  S as createBulletListItemBlockConfig,
  C as createBulletListItemBlockSpec,
  f as createCheckListItemBlockSpec,
  m as createCheckListItemConfig,
  I as createCodeBlockConfig,
  T as createCodeBlockSpec,
  L as createDefaultBlockDOMOutputSpec,
  P as createDividerBlockConfig,
  E as createDividerBlockSpec,
  _ as createFileBlockConfig,
  h as createFileBlockSpec,
  D as createHeadingBlockConfig,
  H as createHeadingBlockSpec,
  b as createImageBlockConfig,
  A as createImageBlockSpec,
  M as createNumberedListItemBlockConfig,
  O as createNumberedListItemBlockSpec,
  N as createPageBreakBlockConfig,
  V as createPageBreakBlockSpec,
  x as createParagraphBlockConfig,
  F as createParagraphBlockSpec,
  G as createQuoteBlockConfig,
  y as createQuoteBlockSpec,
  v as createTableBlockSpec,
  W as createToggleListItemBlockConfig,
  Y as createToggleListItemBlockSpec,
  Q as createToggleWrapper,
  R as createVideoBlockConfig,
  w as createVideoBlockSpec,
  U as defaultBlockSpecs,
  j as defaultBlockToHTML,
  q as defaultInlineContentSchema,
  z as defaultInlineContentSpecs,
  J as defaultProps,
  K as defaultStyleSchema,
  X as defaultStyleSpecs,
  Z as defaultToggledState,
  $ as editorHasBlockWithType,
  ee as fileParse,
  ae as getBackgroundColorAttribute,
  te as getLanguageId,
  ce as getPageBreakSlashMenuItems,
  se as getTextAlignmentAttribute,
  oe as getTextColorAttribute,
  le as imageParse,
  re as imageRender,
  ie as imageToExternalHTML,
  ge as isTableCellSelection,
  ke as mergeParagraphs,
  Be as parseAudioElement,
  ne as parseDefaultProps,
  pe as tablePropSchema,
  de as uploadToTmpFilesDotOrg_DEV_ONLY,
  ue as videoParse,
  Se as withPageBreak
};
